### PR TITLE
Remove reduce and reduceContext from DelayedBucket

### PR DESCRIPTION
**Pull request by iverase**
When creating DelayedBucket, we are passing a reduce and reduceContest to each object which is always the same. In extreme cases we might be creating million of those objects so it seems wasteful to hold those references on those objects.

Therefore this PR proposes to remove them from the object and pass them as paramters on the functions that require them. This might save MiB of memory in extreme cases.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/BucketOrder.java [hunk 1]
```diff
@@ -129,7 +130,10 @@ public final void validate(Aggregator aggregator) throws AggregationExecutionExc
     /**
      * Build a comparator for {@link DelayedBucket}, a wrapper that delays bucket reduction.
      */
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/BucketOrder.java:132]: Probably worth some javadoc warning that the context has to be the one for final reduction. It's pretty obvious to you and me, but I'm thinking of me-in-five-years who will have forgotten lots of things.

File: server/src/main/java/org/elasticsearch/search/aggregations/DelayedBucket.java [hunk 2]
```diff
@@ -36,21 +34,15 @@ public final class DelayedBucket<B extends InternalMultiBucketAggregation.Intern
...
+    public DelayedBucket(List<B> toReduce) {
         this.toReduce = toReduce;
     }
 
     /**
      * The reduced bucket. If the bucket hasn't been reduced already this
      * will reduce the sub-aggs and throw out the list to reduce.
      */
-    public B reduced() {
+    public B reduced(BiFunction<List<B>, AggregationReduceContext, B> reduce, AggregationReduceContext reduceContext) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/DelayedBucket.java:45]: Probably worth the same javadoc.

**Review by nik9000 — APPROVED** [review]
I didn't dig super deeply but it looks like you removed those two members in favor of passing them in where they could be needed and then made the compiler happy.

File: server/src/main/java/org/elasticsearch/search/aggregations/DelayedBucket.java (hunk 2 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/DelayedBucket.java:47]: Added and add an assert on `reduceContext.isFinalReduce()`

File: server/src/main/java/org/elasticsearch/search/aggregations/BucketOrder.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/BucketOrder.java:135]: added

**Comment by iverase** [comment]
>I didn't dig super deeply but it looks like you removed those two members in favor of passing them in where they could be needed and then made the compiler happy.

Very much what I did

**Pull request merged**